### ISSUE TITLE
Request for help with SCSS compile in Joomla

### ISSUE THREAD

**Issue by thednp** [thednp/scss-wasm-compile#1]
Hi @dgrammatiko. How you doin?

As discussed before, we can hangout here on the topic of SCSS compile in Joomla. So I've opted for the in-browser solution of your wasm. I've created a Joomla form field that links sass.js, its worker and a compile script that currently runs `onload` and adds the cassiopeia compiled `template.css` or `template-rtl.css` to the browser console.

In one of your previous posts:

> * Read my comment on the code https://github.com/dgrammatiko/scss-wasm-compile/blob/96e3e44a656dc9f5d5b34ec1847c0fab84fc483f/js.js#L27-L31
> * Implement this missing part on the PHP (regex @import...)
> * implement a simple UI for your user to select a file on the client side
> * for the selected file fetch GET from the PHP the needed SCSS content
> * compile
> * ask the user where to write the compiled css
> * fetch POST the contents and write (overwrite) the file

Some thoughts:

* At this point I'm trying to figure out what to make of your suggestions on `fetch GET from the PHP the needed SCSS content` as well as `fetch POST the contents and write the file`. So far I found [some stuff](https://stackoverflow.com/questions/39414008/fetch-api-get-data-from-php-file) on PHP `fetch`, [stuff](https://stackoverflow.com/questions/64907540/javascript-post-data-to-php-with-fetch) on JavaScript fetch (and others). I also think of [Joomla Input](https://api.joomla.org/cms-3/classes/Joomla.Input.Input.html) which used to be the JRequest. I don't know how to put the compiled CSS back to PHP.
* I don't need to have the user select a source file, it's already known `template.scss` or `template-rtl.scss`.
* As for the destination file, that is also known, it's either `template.min.css` or `template-rtl.min.scss`.

Appreciate any reply.
Cheers

**Comment by dgrammatiko** [comment]
It's dead easy (although the nodesass is deprecated so you need the Dart script)

create a field that appends a label and a button with a unique id and also a script file with `typ=module`

The script is straight forward:
- get the button
- attach a click event listener
- onClick call a function that
- `const response = await fetch('com_ajax...',  {'GET' ... })` This needs a PHP endpoint that will give back the concatenated string of all the SCSS and ALL the imports
- then `const css = await (new Sass()).compile(JSON.parse(response).scss);`
- Then send it back to PHP so it can be saved:  ` await fetch('com_ajax...',  {'PUT' ... })`

**Comment by thednp** [comment]
My field has it's own structure to resolve and handle imports and such, (I can make it create a concatenated SCSS very easy) can I make use of it to create an endpoint for the `com_ajax` part?

Then, I have a namespaced template, with `src` folder, I will create a helper there which will use the field methods to generate the concatenated SCSS content and send it as a response to the `com_ajax`, correct?

Is it possible to interact with the `<button>Save settings</button>` from template style form to do what you've explained? If so, wouldn't it be a better idea?

**Comment by thednp** [comment]
I've created a [new issue](https://joomla.stackexchange.com/questions/28926/use-com-ajax-with-a-custom-field) on StackExchange.

**Comment by dgrammatiko** [comment]
```js

import axios from 'redaxios';

// Get The concatenated SCSS
const getScss = async () => {
  try {
    const response = await axios({
    method: 'get',
    url: '/com_ajax/...', // This needs also the token ?213543254326=1
  });
    const resp = JSON.parse(response);

    return resp.scss
  } catch (error) {
    console.error(error);
  }
}

const compileSCSS = (event) => {
  event.target.setAttribute('disable', '');
  const compileOps = {
    style: Sass.style.expanded,
    precision: -1,
    comments: false,
    indent: '  ',
    linefeed: '\n',
  };

  // Get the SCSS content
  const scssContent = await getScss()

  // compile given SCSS content
  const {css} = await Sass.compile(scssContent, compileOps);

  // Sent the CSS to PHP
  await axios({
    method: 'post',
    url: '/com_ajax/...', // This needs also the token ?213543254326=1
    data: {
      css: css
    }
  });
}

document.getElementById('compileButton').addEventListener('click', compileSCSS);

```

**Comment by thednp** [comment]
@dgrammatiko you mean the PHP code?

**Comment by dgrammatiko** [comment]
Yes the whole thing that you're trying to build. I can't help you without seeing the code

**Comment by thednp** [comment]
@dgrammatiko  https://github.com/thednp/scss-wasm-compile/blob/master/src/Field/CompileField.php

**Comment by dgrammatiko** [comment]
That won't cut it, I need the rest of the files

**Comment by thednp** [comment]
Ok gimme a sec, I will create a new template.

**Comment by thednp** [comment]
@dgrammatiko done.

**Comment by dgrammatiko** [comment]
- Have you tested `concatContent `? It doesn;t seem right, it needs to do things recusively
- You need a Helper.php with the `functionNameAjax()` for the 'GET' (eg this returns the scss text in a json format)
-  You needalso int the HElper.php a  `functionOtherNameAjax()` for the 'Post' (eg this returns will save the data)
- https://docs.joomla.org/Using_Joomla_Ajax_Interface

**Comment by thednp** [comment]
Yes, I have tested the `concatContent` it does exactly what it says it does. What doesn't work is the JS file which is not addapted to work with your suggested changes.

My original `scss-compile.js` file was doing a compile script similar to your [js.js](https://github.com/thednp/scss-wasm-compile/blob/master/js.js) file.

About the `helper.php` file, can it `use TemplateNameSpace\CompileField` and it's classes? If so, how would you write it to work with your above suggested JS please?

**Comment by dgrammatiko** [comment]
> About the helper.php file, can it use TemplateNameSpace\CompileField and it's classes?

NO. Read how com_ajax works: https://github.com/joomla/joomla-cms/blob/4.0-dev/components/com_ajax/ajax.php

**Comment by thednp** [comment]
I've updated my above reply. Now this `ajax.php` is exactly why I don't get it.

**Comment by dgrammatiko** [comment]
Reverse engineer https://github.com/joomla/joomla-cms/tree/4.0-dev/administrator/modules/mod_sampledata
and https://github.com/joomla/joomla-cms/tree/4.0-dev/plugins/sampledata to get a grip how Ajax works. I'm not gonna write the code for you neither I gonna do a free seminar here...

**Comment by thednp** [comment]
So I need a `src/Helper.php` and the JS will have to do something like `index.php?option=com_ajax&template=cassiopeia&method=scssResponseAjax` or `action=scssResponseAjax`.

I don't need a seminar, I need a basic structure I can understand, but that's ok if you think I can take it from here.

**Comment by thednp** [comment]
At this point I understand a bit more:

* the `CompileField.php` will only add the `compile-scss.js` file to the DOM and the `<button>Compile</button>`
* all the methods of the `CompileField.php` will need to get moved to a `cassiopeia/helper.php`
* the JS code you added above will need to use `index.php?option=com_ajax&template=cassiopeia&method=scssResponse&format=json` 
* also that JS will need to use `index.php?option=com_ajax&template=cassiopeia&method=writeScssAjax`.

Click **like** if you agree, I will take some rest and get back to it asap.

Cheers man and thanks again.
